pkg/cli/cli.go

```diff
@@ -0,0 +1,273 @@
+type CLI interface {
+	// Run runs the CLI, usually returning an error if command line configuration
+	// is incorrect.
+	Run() error
+}
+
+// Option is a function that can configure the cli
+type Option func(*cli) error
+
+// cli defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type cli struct {
+	commandName    string
+	defProjVersion string
+	cmd            *cobra.Command
+	extraCommands  []*cobra.Command
+	plugins        map[string][]plugin.Base
+}
+
+// New creates a new cli instance.
+func New(opts ...Option) (CLI, error) {
+	c := &cli{
+		commandName:    "kubebuilder",
+		defProjVersion: internalconfig.DefaultVersion,
+		plugins:        map[string][]plugin.Base{},
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := c.initialize(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run runs the cli.
+func (c cli) Run() error {
+	return c.cmd.Execute()
+}
+
+// WithCommandName is an Option that sets the cli's root command name.
+func WithCommandName(name string) Option {
+	return func(c *cli) error {
+		c.commandName = name
+		return nil
+	}
+}
+
+// WithDefaultProjectVersion is an Option that sets the cli's default project
+// version. Setting an unknown version will result in an error.
+func WithDefaultProjectVersion(version string) Option {
+	return func(c *cli) error {
+		c.defProjVersion = version
+		return nil
+	}
+}
+
+// WithPlugins is an Option that sets the cli's plugins.
+func WithPlugins(plugins ...plugin.Base) Option {
+	return func(c *cli) error {
+		for _, p := range plugins {
+			ver := p.Version()
+			if _, ok := c.plugins[ver]; !ok {
+				c.plugins[ver] = []plugin.Base{}
+			}
+			c.plugins[ver] = append(c.plugins[ver], p)
+		}
+		return nil
+	}
+}
+
+// WithExtraCommands is an Option that adds extra subcommands to the cli.
+// Adding extra commands that duplicate existing commands results in an error.
+func WithExtraCommands(cmds ...*cobra.Command) Option {
+	return func(c *cli) error {
+		c.extraCommands = append(c.extraCommands, cmds...)
+		return nil
+	}
+}
+
+// initialize initializes the cli.
+func (c *cli) initialize() error {
+	var projectVersion string
+	if internal.Configured() {
+		projectVersion, _ = getProjectVersion()
+	} else {
+		// If the project isn't configured, see if we can find the
+		// project version in a command line flag.
+		projectVersion, _ = c.getBaseFlags()
+	}
+
+	if projectVersion != "" {
+		ps, ok := c.plugins[projectVersion]
+		if !ok {
+			return fmt.Errorf("unknown project version %q", projectVersion)
+		}
+		for _, p := range ps {
+			if dp, ok := p.(plugin.Deprecated); ok {
+				fmt.Printf(noticeColor, fmt.Sprintf("[Deprecation Notice] %s\n\n",
+					dp.DeprecationWarning()))
+			}
+		}
+	}
+
+	rootCmd := c.defaultCommand()
+
+	rootCmd.AddCommand(
+		c.newInitProjectCmd(),
+		c.newCreateCmd(),
+	)
+
+	if internal.ConfiguredAndV1() {
```

Comment:
```suggestion
	if internal.IsConfigured && internal.IsV1() {
```
IHMO: Combine 2 or more actions in the same func may hurts concepts as cohesion and single responsibility. Also, I think is easier to read and understanding as suggested.

Reply:
That would require to read the config file twice.